### PR TITLE
do not depend on velocity for approach scaling

### PR DESCRIPTION
**Pull request by Aposhian**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3045  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Not yet |

---

## Description of contribution in a few bullet points
Depend only on the path length left on whether to scale velocity on approach, since depending on lookahead distance results in a cyclic dependency

## Description of documentation updates required from your changes
Added a new parameter.

---

## Future work that may be required in bullet points
Need to test this and build it still.

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/regulated_pure_pursuit_controller.hpp [hunk 1]
```diff
@@ -281,6 +289,7 @@ class RegulatedPurePursuitController : public nav2_core::Controller
   bool use_velocity_scaled_lookahead_dist_;
   tf2::Duration transform_tolerance_;
   double min_approach_linear_velocity_;
+  double approach_velocity_scaling_dist_;
```
Comment (SteveMacenski) [inline nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/regulated_pure_pursuit_controller.hpp:292]: The usual parameter guide + migration guide comment applies :smile:

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 2]
```diff
@@ -325,7 +330,6 @@ geometry_msgs::msg::TwistStamped RegulatedPurePursuitController::computeVelocity
     rotateToHeading(linear_vel, angular_vel, angle_to_heading, speed);
   } else {
     applyConstraints(
```
Comment (SteveMacenski) [inline nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:332]: This call does not contain the `transformed_plan`? Don't believe this compiles

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Beyond the obvious points, this seems good to me!

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp (hunk 2 above)
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:332]: Yes, I haven't fully fleshed this out yet. I will tie up all the loose ends when I get time.

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 3]
```diff
@@ -571,10 +575,39 @@ double RegulatedPurePursuitController::costAtPose(const double & x, const double
   return static_cast<double>(cost);
 }
 
+double RegulatedPurePursuitController::approachVelocityScalingFactor(
+  const nav_msgs::msg::Path::SharedPtr & path
+) const
+{
+  double remaining_distance = nav2_util::geometry_utils::calculate_path_length(path);
+  if (remaining_distance < approach_velocity_scaling_dist_) {
```
[the remark this answers is not in the document]
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:583]: - [ ] document appropriate bounds for approach velocity scaling dist

**Comment by SteveMacenski** [comment]
@Aposhian any update? It would be great to have this in before the next sync & close out https://github.com/ros-planning/navigation2/issues/3045

File: nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/regulated_pure_pursuit_controller.hpp [hunk 4]
```diff
@@ -208,17 +208,25 @@ class RegulatedPurePursuitController : public nav2_core::Controller
    */
   double costAtPose(const double & x, const double & y);
 
+  double approachVelocityScalingFactor(
+    const nav_msgs::msg::Path::SharedPtr & path
```
Comment (padhupradheep) [inline nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/regulated_pure_pursuit_controller.hpp:212]: I guess, we do not want reference to the shared pointer. It does not increase the reference count of the shared pointer.

**Comment by Aposhian** [comment]
I may be able to get to this on Saturday, but I have been a little swamped.

**Comment by Aposhian** [comment]
Here is how it ends up performing with the default 1m. I think that is probably not a good default. I will try to figure out how to set this parameter to have roughly the same behavior as before.
image

**Comment by SteveMacenski** [comment]
Sounds good!

**Comment by Aposhian** [comment]
Ok I just updated the default to be one costmap cell less than the forward costmap extent, so that should approximate the old behavior (where scaling was enabled when lookahead distance was shorter than requested). Here is the behavior with that now on my robot sim:
image
One problem with looking at integrated distance is that it jumps by the robot jumping between poses (this is with a sparse path). However, I think this is a far better behavior than what is existing, and this is also an issue that the velocity smoother could help remedy, correct?

**Comment by SteveMacenski** [comment]
> Ok I just updated the default to be one costmap cell less than the forward costmap extent, so that should approximate the old behavior

Uh, wasn't it before that it started to slow down at the lookahead distance away, e.g. by default about 0.6m? This is much higher than that if its 1/2 the costmap size. The added comment in the readme also doesn't say that you should never set it to be higher than that -- I don't think that should be the default. Its a warning not to exceed, not a recommended value. 

Maybe its a better value, but I suppose you didn't mention that you tried the old setting (~0.6) and found that the longer distance was better. 

Yes, the vel smoother will work out the jaggedness.

**Comment by SteveMacenski** [comment]
> Well, I figured there wasn't a good way to recreate the average lookahead distance, since the velocity is unknown. But maybe the default could be calculated based off lookahead time and desired linear velocity?

Why not just use the value of the parameter `lookahead_dist` as the default? Though, if Adaptive behavior is being used, I suppose just `max_lookahead_dist` would be fine too, but if I had to pick one, I'd pick `lookahead_dist` since that should be no matter what type of PP was being used. Just needs to be documented that it defaults to `lookahead_dist` unless otherwise overridden (and should never exceed costmap size / 2)

**Comment by Aposhian** [comment]
Docs PR: https://github.com/ros-planning/navigation.ros.org/pull/336

**Comment by Aposhian** [comment]
I decided to go with just making it an independent default, but going to 0.6 to match previous behavior. Unit tests are updated. I'm just going to do one more test in simulation with this and then I will be happy with it.

**Comment by Aposhian** [comment]
With some tuning this is what the approach velocity scaling ends up looking like (on a sparse path)
image
I think we can probably do better, but this at least removes the oscillating behavior because of the cyclic dependency.

**Comment by Aposhian** [comment]
@SteveMacenski I think all tasks should be addressed. Let me know if there is anything else.

**Comment by Aposhian** [comment]
If you weren't using adaptive lookahead distance before, then this change will be a regression: it will result in a less smooth scaling. I think we can tweak this algorithm to make it scale linearly, regardless of path density. I just need to brainstorm for a bit.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Toggling for CI, Circle's been flaky lately. Will merge once builds

**Comment by Aposhian** [comment]
Here is what it looks like with the smoothed version:
image

**Pull request merged**